File: api/src/main/kotlin/org/team619/rutor/converter/MainPageGroupedConverter.java
package org.team619.rutor.converter;

import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import org.team619.rutor.core.Converter;
import org.team619.rutor.core.Logger;
import org.team619.rutor.model.Group;
import org.team619.rutor.model.MainGroupedPage;
import org.team619.rutor.model.Row;
import org.team619.rutor.util.Util;

import java.io.Serializable;
import java.util.List;
import java.util.Objects;
import java.util.stream.Collectors;

/**
 * Created by BORIS on 07.08.2016.
 */
public class MainPageGroupedConverter extends RowConverter implements Converter<MainGroupedPage, Document> {

    public MainPageGroupedConverter(Logger logger) {
        super(logger);
    }

    @Override
    public MainGroupedPage convert(Document input) {
        List<Element> elements = input.select(Selectors.GROUP).first().children().stream()
                .skip(1)
                .collect(Collectors.toList());
        List<List<Element>> rawGroups = Util.batch(elements, 2);
        List<Group> groups = rawGroups.stream()
                .map(group -> convertToGroup(group))
                .collect(Collectors.toList());

        return new MainGroupedPage(groups);
    }

    /* =================== */

    private Group convertToGroup(List<Element> group) {
        int index = 0;
        Element header = group.get(index++);
        Element topics = group.get(index++);
        List<Row> rows = topics.select(Selectors.ROWS).stream()
                .skip(1)
                .map(tr -> parseRow(tr.getElementsByTag(Selectors.TD)))
                .filter(Objects::nonNull)
                .collect(Collectors.toList());

        return new Group(header.text(), rows);
    }

    public interface Selectors extends Serializable {

        String GROUP = "#index";
        String ROWS = "tbody > tr";
        String TD = "td";

    }

}

File: api/src/main/kotlin/org/team619/rutor/converter/MainPageGroupedConverter.kt
package org.team619.rutor.converter

import org.jsoup.nodes.Document
import org.jsoup.nodes.Element
import org.team619.rutor.core.Converter
import org.team619.rutor.core.Logger
import org.team619.rutor.model.Group
import org.team619.rutor.model.MainGroupedPage
import org.team619.rutor.util.Constants.Companion.GROUP
import org.team619.rutor.util.Constants.Companion.GROUP_ROWS
import org.team619.rutor.util.Constants.Companion.TD
import org.team619.rutor.util.Util
import java.util.*

/**
 * Created by BORIS on 07.08.2016.
 */
class MainPageGroupedConverter(logger: Logger) : RowConverter(logger), Converter<MainGroupedPage, Document> {

    override fun convert(input: Document): MainGroupedPage {
        val elements = input.select(GROUP)
                .first()
                .children()
                .drop(1)
        val rawGroups = Util.batch(elements, 2)
        val groups = rawGroups
                .map { group -> convertToGroup(group) }

        return MainGroupedPage(groups)
    }

    private fun convertToGroup(group: List<Element>): Group {
        var index = 0
        val header = group[index++]
        val topics = group[index++]
        val rows = topics.select(GROUP_ROWS).drop(1)
                .map { tr ->
                    parseRow(
                            tr.getElementsByTag(TD))
                }
                .filter({ Objects.nonNull(it) })
                .map { it!! }

        return Group(header.text(), rows)
    }

}
